### ISSUE TITLE
Support for REST API based email delivery

### ISSUE THREAD

**Issue by mfkuntz** [TryGhost/Ghost#7930]
### Issue Summary

Ghost can currently send email over SMTP using nodemailer. Adding support to enable sending over API would:

1. Conform to recommended patterns from major mail platforms. ex [Mailgun](https://documentation.mailgun.com/quickstart-sending.html#how-to-start-sending-email)

2. Many hosting providers block SMTP ports by default. There are workarounds, like setting up postfix and proxying on a seperate port, but they are not clean. ex [Google Cloud](https://cloud.google.com/compute/docs/tutorials/sending-mail/)

### Drawbacks

Mail providers don't have a standard API, and there will need to be a custom implementation for each. The official docs reference Mailgun (they seem to be a partner or recommended vendor) so (in my opinion) that would be a good starting point. 

### Requirements

1.  Configuration specification and format. In my testing, there are a couple of standard fields across providers and it should be fairly simple to create a mapping. 

2. Documentation updates explaining the new options and their benefits.

3. Pretty error codes around trying to specify a provider that doesn't exist. 

4. Tests

---

I wrote what I imagine a Mailgun implementation of this would look like, targeted at the stable branch (since that is where I needed it). If this sounds like a good fit for an upstream merge, I can take care of 2 and 3 and submit a pull request.

**Comment by mfkuntz** [comment]
Potential config format (Req#1)

```js
mail: {
    transport: 'API',
    options: {
        service: 'mailgun', // used to lookup a specific implementation
        host: 'mail.example.com', // used if there are multiple domains attached to the mailgun account
        auth: {
            token: 'nope'
        }
    }
},
```

**Comment by kirrg001** [comment]
Hey @mfkuntz  👋

Thanks for your suggestion and interest in working on this. 

Adding new adapters for sending mails requires Ghost to support it. Right now it's [hardcoded](https://github.com/TryGhost/Ghost/blob/master/core/server/api/mail.js#L24) to use the nodemailer integration. 

We already have an issue to make this possible, see https://github.com/TryGhost/Ghost/issues/7687. Helping us making mail as adapter is very welcome 👋 

As soon as Ghost supports installing a different mail adapter, the community is able to implement any mail integration. We already support adapters for storages and scheduling. 

BTW: Last year we already [started with the integration of the Mailgun API](https://github.com/TryGhost/Ghost/pull/6942/files#diff-85ea4b5de9993c7c5d2b8da97eb2d571R1) as part of our existing subscriber feature, which is still in beta.

**Comment by kevinansfield** [comment]
nodemailer already supports different transports (e.g. [this mailgun api transport](https://github.com/orliesaurus/nodemailer-mailgun-transport)) - if self-hosting it may be possible to install this node module and pass the required config params

**Comment by kirrg001** [comment]
@kevinansfield true, but therefor we need to upgrade nodemailer, see https://github.com/TryGhost/Ghost/issues/4125

**Comment by mfkuntz** [comment]
@kirrg001 - Thanks for the adapter link, I will take a look there. 

@kevinansfield - I did take a look at that, but decided not to use it since it seems like adding a dependency directly on a mailgun package is a little weird to put in the core if most people wont use it. Also, I noticed that mailgun doesn't have an official [node library](https://documentation.mailgun.com/libraries.html#node-js), probably because their rest API is so nice.
